### PR TITLE
DATA-891 uri minting tool

### PR DESCRIPTION
**Pull request by markstuart**
Creates a form, model, and migration command to allow logged in users to mint URIs

### CODE REVIEW

File: ckanext/dia/model.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+minted_uri_table = None
+
+
+def db_setup():
+    if minted_uri_table is None:
+        define_table()
+
+    if not model.package_table.exists():
+        log.critical("Exiting: can not migrate minted uri model \
+if the database does not exist yet")
```
Comment (niallmcandrew-work) [inline ckanext/dia/model.py]: Python has implicit string concatenation so you can do this like: 
```suggestion
        log.critical("Exiting: can not migrate minted uri model"
                           "if the database does not exist yet")
```

File: ckanext/dia/model.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        sys.exit(1)
+        return
+
+    if not minted_uri_table.exists():
+        minted_uri_table.create()
+        print("Created Minted URI table")
+    else:
+        print("Minted URI table already exists -- skipping")
```
Comment (niallmcandrew-work) [inline ckanext/dia/model.py]:
```suggestion
        print("Minted URI table already exists -- skipping creation")
```

File: ckanext/dia/model.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    return schema
+
+
+def name_unique(name, context):
+    type = context.get('type', '')
+    if str(type) == '':
+        raise Invalid(_('Type and Name are both required'))
+    # Validate the name is unique within the type
+    result = MintedURI.Session.query(MintedURI).\
+        filter_by(type=type, name=name).first()
```
Comment (niallmcandrew-work) [inline ckanext/dia/model.py]: You can use `exists` here to check if there is an item rather than getting the first.

File: ckanext/dia/model.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+        validated_data, errors = validate(data_dict, default_schema(), data_dict)
+        if errors:
+            raise ValidationError(errors)
+
+        type = validated_data.get('type')
+        name = validated_data.get('name')
+        created_by_id = validated_data.get('created_by_id')
+
+        guid = str(uuid.uuid1())
```
Comment (niallmcandrew-work) [inline ckanext/dia/model.py]: What lead you to UUID v1? I'm wondering if v5 would be more appropriate.

File: ckanext/dia/views.py [hunk 5]
```diff
@@ -2,21 +2,56 @@
...
+            }
+            return tk.render('uris/success.html', extra_vars={'c': extra_vars})
+        except ValidationError as e:
+            vars['errors'] = e.error_dict
+            vars['error_summary'] = e.error_summary
+        except Exception as e:
+            log.error('Unknown error: %s', e, stack_info=True)
+            vars['error_summary'] = { 'Error': _('An unknown error occurred') }
+
+        return tk.render('uris/new.html', extra_vars=vars)
```
Comment (niallmcandrew-work) [inline ckanext/dia/views.py]: Not needed as this will just fall through to the next line.

File: ckanext/dia/views.py [hunk 6]
```diff
@@ -2,21 +2,56 @@
...
-
-
-def get_blueprints():
-    return [dia]
+uri_minter = Blueprint("uri_minter", __name__, template_folder='templates')
+
+@uri_minter.before_request
+def before_request():
+    if not g.userobj:
+        base.abort(403, _(u'Not authorized to see this page'))
```
Comment (niallmcandrew-work) [inline ckanext/dia/views.py:27]: I'd usually recommend a 404 here but given the site is open source it probably doesn't matter. 😆

File: ckanext/dia/model.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    )
+
+
+def default_schema():
+    schema = {
+        'type': [not_empty, six.text_type],
+        'name': [not_empty, six.text_type, name_unique],
+        'created_by_id': [not_empty],
+    }
+    return schema
```
Comment (niallmcandrew-work) [inline ckanext/dia/model.py]: Does this need to be in a function or could it be a constant?

File: ckanext/dia/templates/uris/new.html [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+{% block subtitle %}{{ _('Mint a new URI') }}{% endblock %}
+
+{% block secondary_content %}
+  <section class="module module-narrow module-shallow">
+    <h2 class="module-heading"><i class="fa fa-info-circle"></i> {{ _('URIs for use in linked datasets') }}</h2>
+    <div class="module-content">
+        <p>
+            {% trans %}
+                Some info about what uri minting is and the DIA defined structure
```
Comment (niallmcandrew-work) [inline ckanext/dia/templates/uris/new.html]: Is this to be filled in later?

File: ckanext/dia/views.py [hunk 9]
```diff
@@ -2,21 +2,56 @@
...
+
+@uri_minter.route('/uri/new', methods=['GET', 'POST'])
+def new_uri():
+    vars = {'data': {}, 'errors': {}, 'error_summary': {}}
```
Comment (niallmcandrew-work) [inline ckanext/dia/views.py]: Default the values to `None`, an empty `dict` feels a bit odd when we are assigning to it and not adding items into it.

File: ckanext/dia/views.py [hunk 10]
```diff
@@ -2,21 +2,56 @@
...
+                'type': datatype,
+                'name': name,
+                'created_by_id': g.userobj.id,
+            })
+
+            extra_vars = {
+                'uri': new_instance.uri,
+                'name': new_instance.name,
+            }
+            return tk.render('uris/success.html', extra_vars={'c': extra_vars})
```
Comment (niallmcandrew-work) [inline ckanext/dia/views.py]: Is the `c` key needed?

**Review by niallmcandrew-work — COMMENTED** [review]
A few comments but otherwise looks good to me.

File: ckanext/dia/model.py (hunk 7 above)
Comment (markstuart) [reply ckanext/dia/model.py]: good point, will update to that

File: ckanext/dia/model.py (hunk 4 above)
Comment (markstuart) [reply ckanext/dia/model.py]: I don't really have a strong opinion on this, Judi from Archives NZ/Metataxis suggested an id generated based on machine and current timestamp (which is what had been done in some other prior art she'd been involved with). uuid1 uses that strategy, but I'm not beholden to it. If you have good reason to use a different strat, happy to hear it.

File: ckanext/dia/templates/uris/new.html (hunk 8 above)
Comment (markstuart) [reply ckanext/dia/templates/uris/new.html]: Yes, I was thinking to do that today, likely to get some feedback around this sort of thing once it's on UAT

File: ckanext/dia/views.py (hunk 9 above)
Comment (markstuart) [reply ckanext/dia/views.py]: OK, I'll see how that works with the template on a GET, the fields expect to be populated by data.type and data.name, so unsure if it'll be happy with None.

File: ckanext/dia/views.py (hunk 10 above)
Comment (markstuart) [reply ckanext/dia/views.py]: I guess it'd more in keeping with the pattern here if I set vars['data'] with the info that needs to display in the template... I think this was modelled off some other templates I looked at.

File: ckanext/dia/views.py (hunk 5 above)
Comment (markstuart) [reply ckanext/dia/views.py]: True, will remove

**Review by markstuart — COMMENTED** [review]
Thanks @niall-mediasuite I'll go through and tidy up.

File: ckanext/dia/views.py (hunk 9 above)
Comment (markstuart) [reply ckanext/dia/views.py]: Seems to work ok, will update in next commit

File: ckanext/dia/views.py (hunk 10 above)
Comment (markstuart) [reply ckanext/dia/views.py]: Fixed

File: ckanext/dia/views.py (hunk 5 above)
Comment (markstuart) [reply ckanext/dia/views.py]: Done

File: ckanext/dia/model.py (hunk 4 above)
Comment (markstuart) [reply ckanext/dia/model.py]: uuid5 makes a lot of sense, updated to use this strategy

File: ckanext/dia/model.py (hunk 7 above)
Comment (markstuart) [reply ckanext/dia/model.py]: Function makes more sense with ordering of functions/constants in the file

File: ckanext/dia/model.py (hunk 3 above)
Comment (markstuart) [reply ckanext/dia/model.py]: Didn't seem to work for me, threw an error message when I changed this to `exists()`

**Comment by markstuart** [comment]
image

Listing view with pagination and search over 'type' and 'name'

**Review by niallmcandrew-work — APPROVED** [review]
Looks good to me.

**Pull request merged**